docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,679 @@
+ *         some other handler.
+ *
+ */
+typedef int (*enclave_fault_handler_t)(const sgx_pfinfo *pfinfo, void *private_data);
+
+/* bit 0 - 7 are allocation flags */
+#define EMA_ALLOC_FLAGS_SHIFT 0 
+#define EMA_ALLOC_FLAGS(n) (((unsigned int)(n) << EMA_ALLOC_FLAGS_SHIFT))
+#define EMA_ALLOC_FLAGS_MASK    EMA_ALLOC_FLAGS(0xFF)
+
+/* Only reserve an address range, no physical memory commited.*/
+#define EMA_RESERVE             EMA_ALLOC_FLAGS(1)
+
+/* Reserve address range and commit physical memory. */
+#define EMA_COMMIT_NOW          EMA_ALLOC_FLAGS(2)
+
+/* Reserve address range and commit physical memory on demand.*/
+#define EMA_COMMIT_ON_DEMAND    EMA_ALLOC_FLAGS(4)
+
+/* Always commit pages from higher to lower addresses,
+ *  no gaps in addresses above the last committed.
+ */
+#define EMA_GROWSDOWN           EMA_ALLOC_FLAGS(8)
+
+/* Always commit pages from lower to higher addresses,
+ * no gaps in addresses below the last committed.
+*/
+#define EMA_GROWSUP  EMA_ALLOC_FLAGS(0x10)
+
+/* Map addr must be exactly as requested */     
+#define EMA_FIXED EMA_ALLOC_FLAGS(0x20UL)
+
+/* bit 8 - 15 are page types */
+#define EMA_PAGE_TYPE_SHIFT 8 
+#define EMA_PAGE_TYPE(n) (((unsigned int)(n) << EMA_PAGE_TYPE_SHIFT))
+#define EMA_PAGE_TYPE_MASK      EMA_PAGE_TYPE(0xFF)
+#define EMA_PAGE_TYPE_TCS       EMA_PAGE_TYPE(0x1)  /* TCS page type */
+#define EMA_PAGE_TYPE_REG       EMA_PAGE_TYPE(0x2)  /* regular page type, default if not specified. */
+#define EMA_PAGE_TYPE_TRIM      EMA_PAGE_TYPE(0x4)  /* TRIM page type */
+#define EMA_PAGE_TYPE_SS_FIRST  EMA_PAGE_TYPE(0x5)  /* the first page in shadow stack. */
+#define EMA_PAGE_TYPE_SS_REST   EMA_PAGE_TYPE(0x6)  /* the rest pages in shadow stack. */
+
+/* Use bit 24-32 for alignment masks. */
+#define EMA_ALIGNMENT_SHIFT 24
+/*
+ * Alignment (expressed in log2).  Must be >= log2(PAGE_SIZE) and
+ * < # bits in a pointer (32 or 64).
+ */
+#define EMA_ALIGNED(n) (((unsigned int)(n) << EMA_ALIGNMENT_SHIFT))
+#define EMA_ALIGNMENT_MASK EMA_ALIGNED(0xFFUL)
+#define EMA_ALIGNMENT_64KB EMA_ALIGNED(16UL)
+#define EMA_ALIGNMENT_16MB EMA_ALIGNED(24UL)
+#define EMA_ALIGNMENT_4GB EMA_ALIGNED(32UL)
+
+
+
+/* Permissions flags */
+#define EMA_PROT_NOACCESS   0x0UL
+#define EMA_PROT_READ       0x1UL
+#define EMA_PROT_WRITE      0x2UL
+#define EMA_PROT_EXECUTE    0x4UL
+
+/*
+ * Allocate a new memory region in enclave address space (ELRANGE).
+ * @param[in] addr, starting address of the region, page aligned. If NULL is provided,
+ *                  then the function will select the starting address.
+ * @param[in] length, size of the region in multiples of page size in bytes.
+ * @param[in] flags, a bitwise OR of flags describing committing mode, committing
+ * order, address preference, page type
+ *        Flags should include exactly one of following for committing mode:
+ *            - EMA_RESERVE: just reserve an address range, no EPC commited.
```

Comment:
prefix these defines with SGX_ if they are used with a function (like this one) that starts with sgx_
```suggestion
 *            - SGX_EMA_RESERVE: just reserve an address range, no EPC committed.
```
also note typo at end :)